org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/ChainFinder.java

```diff
@@ -0,0 +1,185 @@
+            }
+        }
+    }
+
+    public void cancel() {
+        isCanceled = true;
+    }
+
+    private void searchChainsForExpectedType(final ChainType expectedType, final int expectedDimensions, final List<ChainElement> entrypoints, final int maxChains, final int minDepth, final int maxDepth) {
+        final LinkedList<LinkedList<ChainElement>> incompleteChains = prepareQueue(entrypoints);
+
+        while (!incompleteChains.isEmpty() && !isCanceled) {
+            final LinkedList<ChainElement> chain = incompleteChains.poll();
+            final ChainElement edge = chain.getLast();
+            if (isValidEndOfChain(edge, expectedType, expectedDimensions)) {
+                if (chain.size() >= minDepth) {
+                    chains.add(new Chain(chain, expectedDimensions));
+                    if (chains.size() == maxChains) {
+                        break;
+                    }
+                }
+                continue;
+            }
+            if (chain.size() < maxDepth && incompleteChains.size() <= 50000) {
+                searchDeeper(chain, incompleteChains, edge.getReturnType());
+            }
+        }
+    }
+
+    /**
+     * Returns the potentially incomplete list of call chains that could be found
+     * before a time out happened. The contents of this list are mutable and may
+     * change as the search makes progress.
+     *
+     * @return The list of call chains
+     */
+    public List<Chain> getChains() {
+        return chains;
+    }
+
+    private static LinkedList<LinkedList<ChainElement>> prepareQueue(final List<ChainElement> entrypoints) {
+        final LinkedList<LinkedList<ChainElement>> incompleteChains = new LinkedList<>();
+        for (final ChainElement entrypoint : entrypoints) {
+            final LinkedList<ChainElement> chain = new LinkedList<>();
+            chain.add(entrypoint);
+            incompleteChains.add(chain);
+        }
+        return incompleteChains;
+    }
+
+    public static boolean isFromExcludedType(final List<String> excluded, final IJavaElement element) {
+        if (element instanceof IType) {
+            return excluded.contains(((IType) element).getFullyQualifiedName());
+        } else {
+            return excluded.contains(element.getElementName());
+        }
+    }
+
+    public static boolean isFromExcludedType(final List<String> excluded, final ChainType element) {
+        if (element.getType() != null) {
+            return isFromExcludedType(excluded, element.getType());
+        }
+        return excluded.contains(element.getPrimitiveType());
+    }
+
+    private boolean isValidEndOfChain(final ChainElement edge, final ChainType expectedType, final int expectedDimension) {
+        if (edge.getElementType() == ElementType.TYPE) {
+            return false;
+        }
+        if (token != null && !token.isBlank() && !edge.getElement().getElementName().startsWith(token)) {
```

Comment:
What if the token is meant to match the start of the chain ? Won't this rule out a bunch of results. Instead of someone trying to match `emptyL` from `Collections.emptyList()`, maybe they type out `Coll` and expect to match from the start of the chain ?

Reply:
I add subword match for both start and end in the latest changes, Let me know what you think.

Reply:
That sounds good.